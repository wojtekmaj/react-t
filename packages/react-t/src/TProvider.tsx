import { useContext, useRef } from 'react';

import TContext from './TContext.js';

import useLocale from './hooks/useLocale.js';

import type { LanguageFiles } from './shared/types.js';

export type TProviderProps<T extends LanguageFiles> = {
  children: React.ReactNode;
  defaultLocale?: Extract<keyof T, string>;
  languageFiles?: LanguageFiles;
  locale?: Extract<keyof T, string>;
  suspend?: boolean;
};

export default function TProvider<T extends LanguageFiles>(props: TProviderProps<T>) {
  const higherTContext = useContext(TContext);

  const { children } = props;
  const {
    defaultLocale,
    languageFiles,
    locale: propsLocale,
    suspend = false,
  } = {
    ...higherTContext,
    ...props,
  };

  const prevSuspend = useRef(suspend);

  if (prevSuspend.current !== suspend) {
    throw new Error('Changing the `suspend` prop of `TProvider` between renders is not supported.');
  }

  const supportedLocales = languageFiles ? Object.keys(languageFiles) : [];

  const locale = useLocale({ defaultLocale, propsLocale, supportedLocales });

  return (
    <TContext.Provider
      value={{
        ...higherTContext,
        languageFiles,
        locale,
        suspend,
      }}
    >
      {children}
    </TContext.Provider>
  );
}
