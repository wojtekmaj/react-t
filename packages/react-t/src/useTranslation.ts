import { cloneElement, isValidElement, useContext, useEffect, useState } from 'react';
import invariant from 'tiny-invariant';

import TContext from './TContext.js';

import use from './use.js';

import { resolveLanguageFile, resolveLanguageFileSync } from './utils/resolver.js';

import type {
  Args,
  ArgsOfNodes,
  ArgsOfStringsOrNumbers,
  GetterOrLanguageFile,
  LanguageFile,
} from './shared/types.js';

function getRawTranslatedString(string: string, languageFile: LanguageFile | null): string {
  return languageFile?.[string] ?? string;
}

function applyArg(
  string: string,
  argName: string,
  replacement: string | number,
  keyPrefix: string,
): string;
function applyArg(
  string: string,
  argName: string,
  replacement: React.ReactNode,
  keyPrefix: string,
): React.ReactNode[];
function applyArg(
  string: string,
  argName: string,
  replacement: string | number | React.ReactNode,
  keyPrefix = '',
): string | React.ReactNode[] {
  const placeholder = `{${argName}}`;
  const regExp = new RegExp(placeholder, 'g');

  if (typeof replacement === 'string' || typeof replacement === 'number') {
    const result = string.replace(regExp, `${replacement}`);

    return result;
  }

  const splitString = string.split(placeholder);

  const initialValue: React.ReactNode[] = [];

  return splitString.reduce((arr, element, index) => {
    const isLast = index === splitString.length - 1;

    arr.push(element);

    if (!isLast) {
      arr.push(
        isValidElement(replacement)
          ? // biome-ignore lint/suspicious/noArrayIndexKey: index is stable here
            cloneElement(replacement, { key: `${keyPrefix}-${index}` })
          : replacement,
      );
    }

    return arr;
  }, initialValue);
}

function applyArgs(string: string, args: ArgsOfStringsOrNumbers): string;
function applyArgs(string: string, args: ArgsOfNodes): React.ReactNode[];
function applyArgs(rawString: string, args: Args): string | React.ReactNode[] {
  let result: React.ReactNode[] = [rawString];

  for (const [argName, replacement] of Object.entries(args || {})) {
    result = result.flatMap((string, index) => {
      return typeof string === 'string'
        ? applyArg(string, argName, replacement, `${index}`)
        : string;
    });
  }

  return result;
}

const resolveCache = new Map<GetterOrLanguageFile, Promise<LanguageFile | null>>();

function useLanguageFileSuspense(getterOrLanguageFile: GetterOrLanguageFile | null) {
  if (!getterOrLanguageFile) {
    return null;
  }

  const languageFileSync = resolveLanguageFileSync(getterOrLanguageFile);

  if (languageFileSync) {
    return languageFileSync;
  }

  const promise = (() => {
    if (!resolveCache.has(getterOrLanguageFile)) {
      resolveCache.set(getterOrLanguageFile, resolveLanguageFile(getterOrLanguageFile));
    }

    const cachedPromise = resolveCache.get(getterOrLanguageFile);

    if (!cachedPromise) {
      throw new Error('Unable to find cached promise');
    }

    return cachedPromise;
  })();

  const languageFileAsync = use(promise);

  return languageFileAsync;
}

function useLanguageFileState(getterOrLanguageFile: GetterOrLanguageFile | null) {
  const [languageFile, setLanguageFile] = useState<LanguageFile | null>(
    resolveLanguageFileSync(getterOrLanguageFile),
  );

  useEffect(() => {
    resolveLanguageFile(getterOrLanguageFile).then(setLanguageFile);
  }, [getterOrLanguageFile]);

  return languageFile;
}

export default function useTranslation(string?: undefined, args?: undefined): undefined;
export default function useTranslation(string: undefined, args?: Args): undefined;
export default function useTranslation<T extends string | undefined>(
  string: T,
  args?: undefined,
): T extends string ? string : undefined;
export default function useTranslation<T extends string | undefined>(
  string: T,
  args: ArgsOfStringsOrNumbers,
): T extends string ? string : undefined;
export default function useTranslation<T extends string | undefined>(
  string: T,
  args: ArgsOfNodes,
): T extends string ? React.ReactNode[] : undefined;
export default function useTranslation(
  string?: string,
  args?: Args,
): string | React.ReactNode[] | undefined {
  const context = useContext(TContext);

  invariant(context, 'Unable to find TProvider context. Did you wrap your app in <TProvider />?');

  const { languageFiles, locale, suspend } = context;

  const getterOrLanguageFile =
    (languageFiles && locale && languageFiles[locale] ? languageFiles[locale] : null) || null;

  // We're breaking the rules of hooks here because we know that suspend will never change
  const languageFile = suspend
    ? useLanguageFileSuspense(getterOrLanguageFile)
    : useLanguageFileState(getterOrLanguageFile);

  if (!string) {
    return string;
  }

  const rawTranslatedString = getRawTranslatedString(string, languageFile);

  if (!args) {
    return rawTranslatedString;
  }

  const stringWithArgs = applyArgs(rawTranslatedString, args);

  return stringWithArgs;
}
