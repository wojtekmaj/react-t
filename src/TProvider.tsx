import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import TContext from './TContext.js';

import { resolveLanguageFile, resolveLanguageFileSync } from './utils/resolver.js';

import useLocale from './hooks/useLocale.js';

import { LanguageFile, LanguageFiles } from './shared/types.js';

export type TProviderProps<T extends LanguageFiles> = {
  children: React.ReactNode;
  defaultLocale?: Extract<keyof T, string>;
  languageFiles?: LanguageFiles;
  locale?: Extract<keyof T, string>;
};

const TProvider: React.FC<TProviderProps<LanguageFiles>> = function TProvider<
  T extends LanguageFiles,
>({ children, defaultLocale, languageFiles, locale: propsLocale }: TProviderProps<T>) {
  const supportedLocales = languageFiles ? Object.keys(languageFiles) : [];

  const locale = useLocale({ defaultLocale, propsLocale, supportedLocales });

  const getterOrLanguageFile =
    languageFiles && locale && languageFiles[locale] ? languageFiles[locale] : undefined;

  const [languageFile, setLanguageFile] = useState<LanguageFile | undefined>(
    resolveLanguageFileSync(getterOrLanguageFile),
  );

  useEffect(() => {
    resolveLanguageFile(getterOrLanguageFile).then(setLanguageFile);
  }, [getterOrLanguageFile]);

  return <TContext.Provider value={{ languageFile }}>{children}</TContext.Provider>;
};

TProvider.propTypes = {
  children: PropTypes.node,
  defaultLocale: PropTypes.string,
  languageFiles: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.any, PropTypes.func])),
  locale: PropTypes.string,
};

export default TProvider;
