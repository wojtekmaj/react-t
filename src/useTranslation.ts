import { cloneElement, isValidElement, useContext } from 'react';
import invariant from 'tiny-invariant';

import TContext from './TContext.js';

import type { Args, ArgsOfNodes, ArgsOfStringsOrNumbers, LanguageFile } from './shared/types.js';

function getRawTranslatedString(string: string, languageFile: LanguageFile | undefined): string {
  return languageFile?.[string] ?? string;
}

function applyArg(
  string: string,
  argName: string,
  replacement: string | number,
  keyPrefix: string,
): string;
function applyArg(
  string: string,
  argName: string,
  replacement: React.ReactNode,
  keyPrefix: string,
): React.ReactNode[];
function applyArg(
  string: string,
  argName: string,
  replacement: string | number | React.ReactNode,
  keyPrefix = '',
): string | React.ReactNode[] {
  const placeholder = `{${argName}}`;
  const regExp = new RegExp(placeholder, 'g');

  if (typeof replacement === 'string' || typeof replacement === 'number') {
    const result = string.replace(regExp, `${replacement}`);

    return result;
  }

  const splitString = string.split(placeholder);

  const initialValue: React.ReactNode[] = [];

  return splitString.reduce((arr, element, index) => {
    const isLast = index === splitString.length - 1;

    if (isLast) {
      return [...arr, element];
    }

    return [
      ...arr,
      element,
      isValidElement(replacement)
        ? // eslint-disable-next-line react/no-array-index-key
          cloneElement(replacement, { key: `${keyPrefix}-${index}` })
        : replacement,
    ];
  }, initialValue);
}

function applyArgs(string: string, args: ArgsOfStringsOrNumbers): string;
function applyArgs(string: string, args: ArgsOfNodes): React.ReactNode[];
function applyArgs(rawString: string, args: Args): string | React.ReactNode[] {
  let result: React.ReactNode[] = [rawString];

  for (const [argName, replacement] of Object.entries(args || {})) {
    result = result.flatMap((string, index) => {
      return typeof string === 'string'
        ? applyArg(string, argName, replacement, `${index}`)
        : string;
    });
  }

  return result;
}

export default function useTranslation(string?: undefined, args?: undefined): undefined;
export default function useTranslation(string: undefined, args?: Args): undefined;
export default function useTranslation<T extends string | undefined>(
  string: T,
  args?: undefined,
): T extends string ? string : undefined;
export default function useTranslation<T extends string | undefined>(
  string: T,
  args: ArgsOfStringsOrNumbers,
): T extends string ? string : undefined;
export default function useTranslation<T extends string | undefined>(
  string: T,
  args: ArgsOfNodes,
): T extends string ? React.ReactNode[] : undefined;
export default function useTranslation(
  string?: string,
  args?: Args,
): string | React.ReactNode[] | undefined {
  const context = useContext(TContext);

  invariant(context, 'Unable to find TProvider context. Did you wrap your app in <TProvider />?');

  const { languageFile } = context;

  if (!string) {
    return string;
  }

  const rawTranslatedString = getRawTranslatedString(string, languageFile);

  if (!args) {
    return rawTranslatedString;
  }

  const stringWithArgs = applyArgs(rawTranslatedString, args);

  return stringWithArgs;
}
