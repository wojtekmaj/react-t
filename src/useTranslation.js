import { useContext, useMemo } from 'react';
import invariant from 'tiny-invariant';

import { Context } from './TProvider';

function getTranslatedString(languageFile, string) {
  if (languageFile && typeof languageFile[string] === 'string') {
    return languageFile[string];
  }

  return string;
}

function applyVars(rawString, args) {
  if (!args) {
    return rawString;
  }

  let finalString = rawString;
  Object.entries(args).forEach(([key, value]) => {
    finalString = finalString.replace(`{${key}}`, value);
  });

  return finalString;
}

export default function useTranslation(string, args) {
  const context = useContext(Context);

  invariant(context, 'Unable to find TProvider context. Did you wrap your app in <TProvider />?');

  const { languageFile } = context;

  const translatedString = useMemo(() => {
    const rawTranslatedString = getTranslatedString(languageFile, string);
    const stringWithArgs = applyVars(rawTranslatedString, args);

    return stringWithArgs;
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [languageFile, string, JSON.stringify(args)]);

  return translatedString;
}
