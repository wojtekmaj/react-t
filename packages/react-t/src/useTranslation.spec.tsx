import { beforeEach, describe, expect, it, vi } from 'vitest';
import { act, renderHook, waitFor } from '@testing-library/react';
import { getUserLocales } from 'get-user-locale';

import useTranslation from './useTranslation.js';
import TProvider from './TProvider.js';

import { muteConsole, restoreConsole } from '../../../test-utils.js';

import type { LanguageFile, LanguageFileModule } from './shared/types.js';

const deLanguageFile: LanguageFile = {
  'Hello world!': 'Hallo Welt!',
  'Hello {name}!': 'Hallo {name}!',
  'Hello {name} and {other}!': 'Hallo {name} und {other}!',
  'Hello {name}! Nice to meet you {name}!': 'Hallo {name}! Schön, dich zu treffen {name}!',
};

const esLanguageFile: LanguageFile = {
  'Hello world!': '¡Hola Mundo!',
  'Hello {name}!': '¡Hola {name}!',
  'Hello {name} and {other}!': '¡Hola {name} y {other}!',
  'Hello {name}! Nice to meet you {name}!': '¡Hola {name}! ¡Encantado de conocerte {name}!',
};

const languageFiles = {
  'de-DE': deLanguageFile,
  'es-ES': esLanguageFile,
};

const syncLanguageFiles = {
  'de-DE': () => deLanguageFile,
  'es-ES': () => esLanguageFile,
};

const asyncLanguageFiles: Record<string, () => Promise<LanguageFile>> = {
  'de-DE': () => new Promise((resolve) => resolve(deLanguageFile)),
  'es-ES': () => new Promise((resolve) => resolve(esLanguageFile)),
};

const asyncLanguageFilesEsm: Record<string, () => Promise<LanguageFileModule>> = {
  'de-DE': () => new Promise((resolve) => resolve({ default: deLanguageFile })),
  'es-ES': () => new Promise((resolve) => resolve({ default: esLanguageFile })),
};

vi.mock('get-user-locale', () => ({
  getUserLocales: vi.fn(),
}));

function uniq<T>(array: T[]): T[] {
  return Array.from(new Set(array));
}

vi.mocked(getUserLocales).mockImplementation(() =>
  uniq([...navigator.languages, navigator.language, 'en-US']),
);

describe('useTranslation() hook', () => {
  beforeEach(() => {
    document.documentElement.removeAttribute('lang');
  });

  it('throws when rendered without TProvider context', () => {
    muteConsole();

    expect(() => renderHook(() => useTranslation('Hello world!'))).toThrowError();

    restoreConsole();
  });

  it('renders nothing given nothing', () => {
    const { result } = renderHook(() => useTranslation(), { wrapper: TProvider });

    expect(result.current).toBe(undefined);
  });

  it('returns original phrase given no language files', () => {
    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper: TProvider });

    expect(result.current).toBe('Hello world!');
  });

  it('returns original phrase with variable given no language files', () => {
    const { result } = renderHook(() => useTranslation('Hello {name}!', { name: 'John' }), {
      wrapper: TProvider,
    });

    expect(result.current).toBe('Hello John!');
  });

  it('returns original phrase with ReactNode variable given no language files', () => {
    const { result } = renderHook(
      () => useTranslation('Hello {name}!', { name: <strong>John</strong> }),
      { wrapper: TProvider },
    );

    expect(result.current).toMatchSnapshot();
  });

  it('returns original phrase with variable placeholder given no language files and no args', () => {
    const { result } = renderHook(() => useTranslation('Hello {name}!'), { wrapper: TProvider });

    expect(result.current).toBe('Hello {name}!');
  });

  it('returns original phrase with multiple variables given no language files', () => {
    const { result } = renderHook(
      () => useTranslation('Hello {name} and {other}!', { name: 'John', other: 'Elisabeth' }),
      { wrapper: TProvider },
    );

    expect(result.current).toBe('Hello John and Elisabeth!');
  });

  it('returns original phrase with one variable used multiple times given no language files', () => {
    const { result } = renderHook(
      () => useTranslation('Hello {name}! Nice to meet you {name}!', { name: 'John' }),
      { wrapper: TProvider },
    );

    expect(result.current).toBe('Hello John! Nice to meet you John!');
  });

  it('returns original phrase if html lang is given but no languageFiles were given', () => {
    muteConsole();

    document.documentElement.setAttribute('lang', 'de-DE');

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper: TProvider });

    expect(result.current).toBe('Hello world!');

    restoreConsole();
  });

  it('returns original phrase if html lang equal to defaultLanguage is given', async () => {
    document.documentElement.setAttribute('lang', 'en-US');

    const { result } = renderHook(() => useTranslation('Hello world!'), {
      wrapper: TProvider,
    });

    expect(result.current).toBe('Hello world!');
  });

  it('returns original phrase if language file is still loading', async () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    const { result } = renderHook(() => useTranslation('Hello world!'), {
      wrapper: TProvider,
    });

    expect(result.current).toBe('Hello world!');
  });

  it('returns translated phrase if html lang is given and language files are given', async () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={languageFiles}>{children}</TProvider>;
    }

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper });

    expect(result.current).toBe('Hallo Welt!');
  });

  it('returns translated phrase with variable if locale prop is given', () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return (
        <TProvider languageFiles={languageFiles} locale="de-DE">
          {children}
        </TProvider>
      );
    }

    const { result } = renderHook(() => useTranslation('Hello {name}!', { name: 'John' }), {
      wrapper,
    });

    expect(result.current).toBe('Hallo John!');
  });

  it('returns translated phrases with lang overwritten by the second TProvider', async () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return (
        <TProvider languageFiles={languageFiles}>
          <TProvider locale="es-ES">{children}</TProvider>
        </TProvider>
      );
    }

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper });

    expect(result.current).toBe('¡Hola Mundo!');
  });

  it('returns translated phrase with ReactNode variable if locale prop is given', () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return (
        <TProvider languageFiles={languageFiles} locale="de-DE">
          {children}
        </TProvider>
      );
    }

    const { result } = renderHook(
      () => useTranslation('Hello {name}!', { name: <strong>John</strong> }),
      { wrapper },
    );

    expect(result.current).toMatchSnapshot();
  });

  it('returns translated phrase with variable placeholder if locale prop is given given no args', () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={languageFiles}>{children}</TProvider>;
    }

    const { result } = renderHook(() => useTranslation('Hello {name}!'), { wrapper });

    expect(result.current).toBe('Hallo {name}!');
  });

  it('returns translated phrase with multiple variables if locale prop is given', () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return (
        <TProvider languageFiles={languageFiles} locale="de-DE">
          {children}
        </TProvider>
      );
    }

    const { result } = renderHook(
      () => useTranslation('Hello {name} and {other}!', { name: 'John', other: 'Elisabeth' }),
      { wrapper },
    );

    expect(result.current).toBe('Hallo John und Elisabeth!');
  });

  it('returns translated phrase with one variable used multiple times if locale prop is given', () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={languageFiles}>{children}</TProvider>;
    }

    const { result } = renderHook(
      () => useTranslation('Hello {name}! Nice to meet you {name}!', { name: 'John' }),
      { wrapper },
    );

    expect(result.current).toBe('Hallo John! Schön, dich zu treffen John!');
  });

  it('returns translated phrase if html lang is given and synchronous functions returning language files are given', async () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={syncLanguageFiles}>{children}</TProvider>;
    }

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper });

    expect(result.current).toBe('Hallo Welt!');
  });

  it('returns translated phrase if html lang is given and asynchronous functions returning language files are given', async () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={asyncLanguageFiles}>{children}</TProvider>;
    }

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper });

    await waitFor(() => expect(result.current).toBe('Hallo Welt!'));
  });

  it('returns translated phrase if html lang is given and asynchronous functions returning language files as ESM modules are given', async () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={asyncLanguageFilesEsm}>{children}</TProvider>;
    }

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper });

    await waitFor(() => expect(result.current).toBe('Hallo Welt!'));
  });

  it('changes translated phrase if html lang is changed', async () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={asyncLanguageFiles}>{children}</TProvider>;
    }

    const { rerender, result } = renderHook(() => useTranslation('Hello world!'), {
      wrapper,
    });

    await waitFor(() => expect(result.current).toBe('Hallo Welt!'));

    act(() => {
      document.documentElement.setAttribute('lang', 'es-ES');
    });

    rerender();

    await waitFor(() => expect(result.current).toBe('¡Hola Mundo!'));
  });

  it('changes translated phrase if html lang is changed to value equal to defaultLanguage', async () => {
    document.documentElement.setAttribute('lang', 'de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={asyncLanguageFiles}>{children}</TProvider>;
    }

    const { rerender, result } = renderHook(() => useTranslation('Hello world!'), {
      wrapper,
    });

    await waitFor(() => expect(result.current).toBe('Hallo Welt!'));

    act(() => {
      document.documentElement.setAttribute('lang', 'en-US');
    });

    rerender();

    await waitFor(() => expect(result.current).toBe('Hello world!'));
  });

  it('returns original phrase if browser language is given but no languageFiles were given', () => {
    muteConsole();

    const languageGetter = vi.spyOn(window.navigator, 'language', 'get');
    languageGetter.mockReturnValue('de-DE');

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper: TProvider });

    expect(result.current).toBe('Hello world!');

    languageGetter.mockRestore();

    restoreConsole();
  });

  it('returns translated phrase if browser language is given', async () => {
    const languageGetter = vi.spyOn(window.navigator, 'language', 'get');
    languageGetter.mockReturnValue('de-DE');

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider languageFiles={asyncLanguageFiles}>{children}</TProvider>;
    }

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper });

    await waitFor(() => expect(result.current).toBe('Hallo Welt!'));

    languageGetter.mockRestore();
  });

  it('returns original phrase if locale prop is given but no languageFiles were given', () => {
    muteConsole();

    function wrapper({ children }: { children: React.ReactNode }) {
      return <TProvider locale="de-DE">{children}</TProvider>;
    }

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper });

    expect(result.current).toBe('Hello world!');

    restoreConsole();
  });

  it('returns translated phrase if locale prop is given', async () => {
    function wrapper({ children }: { children: React.ReactNode }) {
      return (
        <TProvider languageFiles={asyncLanguageFiles} locale="de-DE">
          {children}
        </TProvider>
      );
    }

    const { result } = renderHook(() => useTranslation('Hello world!'), { wrapper });

    await waitFor(() => expect(result.current).toBe('Hallo Welt!'));
  });

  it('returns empty string if the translated phrase is an empty string', async () => {
    function wrapper({ children }: { children: React.ReactNode }) {
      return (
        <TProvider languageFiles={{ 'de-DE': { foo: 'foo', bar: '' } }} locale="de-DE">
          {children}
        </TProvider>
      );
    }

    const { result } = renderHook(() => useTranslation('bar'), { wrapper });

    await waitFor(() => expect(result.current).toBe(''));
  });
});
