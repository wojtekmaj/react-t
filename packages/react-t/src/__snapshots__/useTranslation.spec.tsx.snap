// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useTranslation() hook > returns original phrase with ReactNode variable given no language files 1`] = `
[
  "Hello ",
  <strong>
    John
  </strong>,
  "!",
]
`;

exports[`useTranslation() hook > returns translated phrase with ReactNode variable if locale prop is given 1`] = `
[
  "Hallo ",
  <strong>
    John
  </strong>,
  "!",
]
`;
